vavr/src/main/java/io/vavr/collection/RedBlackTree.java

```diff
@@ -78,14 +79,101 @@
+            if (comparator.compare(lastItem, values[i]) != 0) {
+                lastIndex++;
+            }
+
+            values[lastIndex] = values[i];
+            lastItem = values[i];
+        }
+
+        return ofSortedAndDeduplicatedArray(comparator, values, lastIndex + 1);
+    }
+
+    static <T> RedBlackTree<T> ofSortedAndDeduplicatedArray(Comparator<? super T> comparator, T values[], int length) {
+        final Empty<T> empty = new Empty<>(comparator);
+
+        if (length == 0) {
+            return empty;
+        }
+
+        int height = length == 1 ? 1 : log2IntRoundedDown(length);
+
+        return ofSortedAndDeduplicatedArray(empty, values, height, 0, length);
+    }
+
+    static int log2IntRoundedDown(int count) {
+        return (Integer.SIZE - Integer.numberOfLeadingZeros(count)) - 1;
+    }
+
+    // Based on https://www.cs.princeton.edu/~rs/talks/LLRB/RedBlack.pdf
+    // We construct 2-nodes all the way down, inserting 3 and 4-nodes at the lowest level to fill in the remaining items but keep the blackHeight invariant
+    static <T> RedBlackTree<T> ofSortedAndDeduplicatedArray(Empty<T> empty, T values[], int remainingHeight, int start, int size) {
+        if (size == 0) {
+            if (remainingHeight != 0) {
+                throw new IllegalStateException("Expected remainigHeight = 0 for leaf");
+            }
+            // Make a leaf.
+            return empty;
+        } else if (size == 1) {
+            if (remainingHeight != 1) {
+                throw new IllegalStateException("Expected remainigHeight = 1 for bottom 2-node, got " + remainingHeight);
+            }
+            // Make a 2-node
+            // Invariant: remainingHeight should be 1
+            return new Node<T>(BLACK, remainingHeight, empty, values[start], empty, empty);
+        } else if (size == 2 && remainingHeight == 1) {
+            // Make a 3-node, to fill the remainder
+            return new Node<T>(BLACK, remainingHeight, new Node<T>(RED, remainingHeight, empty, values[start], empty, empty), values[start + 1], empty, empty);
+        } else if (size == 3 && remainingHeight == 1) {
+            // Make a 4-node, to fill the remainder
+            return new Node<T>(BLACK, remainingHeight,
+                    new Node<T>(RED, remainingHeight, empty, values[start], empty, empty),
+                    values[start + 1],
+                    new Node<T>(RED, remainingHeight, empty, values[start + 2], empty, empty),
+                    empty);
+        } else {
+            // Make a 2-node and recurse
+            int childCount = size - 1;
+            int perChild = childCount / 2;
+            int remainder = childCount % 2;// Make a 4-node and recurse
+            int child1Count = perChild + (remainder >= 1 ? 1 : 0);
+            int child2Count = perChild;
+
+            RedBlackTree<T> c1 = ofSortedAndDeduplicatedArray(empty, values, remainingHeight - 1, start, child1Count);
```

Comment:
The recursion here seems like it might Stack overflow when used on a (very) large input.

Reply:
log2 should keep the call depth to a minimum